### PR TITLE
FEAT: On-the-fly benchmarking + import timing improvements :clock1:  v1

### PR DESCRIPTION
**Pull request by daquinteroflex**
See further details https://github.com/flexcompute/tidy3d/pull/1353

Now we have a new timing utility included in this PR and we can use it on different pieces of hardware

```
poetry run tidy3d develop benchmark-timing-operations -c average_test_import
```

This version contains a:
* Generally faster import
* No vtk restructure

EDIT: in order to solve the last remaining issues, this PR now also contains:
* FAQ submodule additions
* `tidy3d develop` cli testing
* Improved further documentation with myST
* Reimplemented the old parameter visualiser issue

### CODE REVIEW

**Comment by daquinteroflex** [comment]
For reference this version is this good:
```
average tidy3d load time = 1.0373110938072205
```

image

**Comment by daquinteroflex** [comment]
Note that this is the same as https://github.com/flexcompute/tidy3d/pull/1376 just that with the benchmark utility for pre/2.6

File: tidy3d/components/types_extra.py
Comment (tylerflex) [reply tidy3d/components/types_extra.py:1]: also, a few of the other new files are missing them as well.

File: tidy3d/web/cli/develop/packaging.py
Comment (tylerflex) [inline tidy3d/web/cli/develop/packaging.py:1]: ss there any mention of these new features in the development guide?

File: tidy3d/web/cli/develop/packaging.py
Comment (tylerflex) [inline tidy3d/web/cli/develop/packaging.py:44]: do we need `pass`?

File: tidy3d/web/cli/develop/packaging.py
Comment (tylerflex) [inline tidy3d/web/cli/develop/packaging.py:62]: could probably use single quotes around `timing_command` and a space after. (or could use an `f` string).

**Review by tylerflex — CHANGES_REQUESTED** [review]
Thanks @daquinteroflex I had a read through, there were some minor comments but overall looks good.

I think the `verify_packages_import(["trimesh"])` approach is nice, seems to simplify things a bit.

I'm a little confused why vtk is separated into `tidy3d/packaging.py` and trimesh and the others seem to be in types/types extra? Is there a reason these weren't all put together?

The new develop CLI functions are a bit hard to follow. I'm not exactly sure where they're used but also didn't go into a ton of detail. If you can make sure to future proof these by adding more discussion to the developer guide, that would probably go a long way to making sure it can be fixed or extended.

Also, is there currently any notion of testing the develop CLI? I wonder if we should consider this (like a test that uses the tidy3d CLI to set everything up and run through all the commands?) Sorry if this already exists but I missed it.

File: tests/test_timing/import_tidy3d.py
Comment (daquinteroflex) [reply tests/test_timing/import_tidy3d.py:8]: Sounds good, yeah might be a bit more reliable. I'll test between them as it's just a few line changes. In either case, it was just to mimic the package unload process.

File: tidy3d/components/geometry/base.py
Comment (daquinteroflex) [reply tidy3d/components/geometry/base.py:1092]: Will be fixed in https://github.com/flexcompute/tidy3d/pull/1376 and propagated through.

File: tidy3d/components/types_extra.py
Comment (daquinteroflex) [reply tidy3d/components/types_extra.py:1]: So it's actually intentional. I will comment about this since it relates to Momchil's comment in https://github.com/flexcompute/tidy3d/pull/1376

File: tidy3d/packaging.py
Comment (daquinteroflex) [reply tidy3d/packaging.py:107]: Yeah sorry was conceptual but will be fixed in https://github.com/flexcompute/tidy3d/pull/1376 based on Lucas's logic recommendation and propagated through

File: tidy3d/packaging.py
Comment (daquinteroflex) [reply tidy3d/packaging.py:90]: Yeah can do, so the idea was that we can also extend this function to provide an package import state array which could come handy in some future and thought it was more verbose but each case had just specific operations, without repeating the iteration operations.

I intend to modify it to Lucas's and your suggestion

File: tidy3d/packaging.py
Comment (daquinteroflex) [reply tidy3d/packaging.py:107]: Yeah sorry was conceptual but will be fixed in https://github.com/flexcompute/tidy3d/pull/1376 based on Lucas's logic recommendation and propagated through

File: tidy3d/packaging.py
Comment (daquinteroflex) [reply tidy3d/packaging.py:90]: Actually between both of your suggestions we can do something more operationally efficient so am implementing fully

File: tests/test_timing/import_tidy3d.py
Comment (daquinteroflex) [reply tests/test_timing/import_tidy3d.py:8]: So I did an attempt at this and the problem is the unloading of the package, it doesn't really do it with `timeit`, so have found to need to use this script which I think is representative to my own terminal experience.

File: tidy3d/web/cli/develop/__init__.py
Comment (daquinteroflex) [reply tidy3d/web/cli/develop/__init__.py:1]: Sorted

File: tidy3d/web/cli/develop/packaging.py
Comment (daquinteroflex) [reply tidy3d/web/cli/develop/packaging.py:62]: Sorted

File: tidy3d/web/cli/develop/packaging.py
Comment (daquinteroflex) [reply tidy3d/web/cli/develop/packaging.py:44]: Dev leftovers sorted

File: tidy3d/web/cli/develop/packaging.py
Comment (daquinteroflex) [reply tidy3d/web/cli/develop/packaging.py:1]: Now they're in better

File: tidy3d/web/cli/develop/documentation.py
Comment (daquinteroflex) [reply tidy3d/web/cli/develop/documentation.py:89]: Sorted

File: tidy3d/components/geometry/base.py
Comment (daquinteroflex) [reply tidy3d/components/geometry/base.py:1292]: Wow that's a typo, good find. Didn't error.

Comment (daquinteroflex) [reply tidy3d/components/geometry/base.py:1292]: Fixed now but weird

Comment (daquinteroflex) [reply tidy3d/components/geometry/base.py:1292]: According to the tests that just failed it was not a typo

**Comment by daquinteroflex** [comment]
Consolidated various PRs and changes in here, and I think this needs another review in case you can @tylerflex . Tests basically about to pass and minimum changes expected

**Comment by daquinteroflex** [comment]
Wohoo tests are passing and it's ready to be reviewed again and rebased!

File: tests/test_cli/full_test_develop.py
Comment (tylerflex) [reply tests/test_cli/full_test_develop.py:98]: this applies to the other tests in test_packaging and others.

File: tidy3d/components/geometry/polyslab.py
Comment (tylerflex) [inline tidy3d/components/geometry/polyslab.py:556]: I was dealing with a similar thing in my 3.0 architecture exercise yesterday. I was wondering if there's any way to "inject" this package into the function scope namespace, such that if `verify_packages_import("x")` works, then `x` is imported in the function's scope (and we dont need to import it manually). sounds like it could be potentially buggy but just a thought.

File: tidy3d/components/types_extra.py
Comment (tylerflex) [reply tidy3d/components/types_extra.py:1]: thanks, still would be good to put a module level docstring though (I thought ruff would have caught these)

File: tidy3d/web/cli/develop/documentation.py
Comment (tylerflex) [inline tidy3d/web/cli/develop/documentation.py:1]: so were these new files in `web/cli/develop` just moved from formerly `web/cli/develop.py` and then just split up? that is to say, these aren't new functions, just moved?

File: tests/test_cli/full_test_develop.py
Comment (tylerflex) [reply tests/test_cli/full_test_develop.py:1]: If we run test coverage including all of the CLI stuff, what are the % before and after? do you know? Also, what % of the develop commands does it hit?

File: .github/workflows/test_develop_cli.yaml
Comment (daquinteroflex) [reply .github/workflows/test_develop_cli.yaml:10]: So I was thinking just triggered on the `develop`  PR or manually, how does that sound?

The nice thing about this script is that it tests that the docs build, unsure how to test for a release since it is just creating a tag that triggers that? Could add a pre trigger run if we wanted on any release?

File: docs/development/documentation.rst
Comment (daquinteroflex) [reply docs/development/documentation.rst:49]: So actually I made this for the web api because it changed between 2.5 and 2.6. I then realised that it was backward compatible after I had made this :facepalm: . Could come handy from 2 -> 3 most likely to be honest and in general beneficial to have this since it can update all docs throughout all repos submoduled.

File: tidy3d/components/base.py
Comment (daquinteroflex) [reply tidy3d/components/base.py:1]: Exactly, the docs build on the parameters looks like before.

File: .github/workflows/run_tests.yml
Comment (daquinteroflex) [reply .github/workflows/run_tests.yml:54]: I found some errors before when it had 1.4.1 and tend to test locally, so I found it handy to have checked before locally before it upgrades personally.

File: tidy3d/web/cli/develop/documentation.py
Comment (daquinteroflex) [reply tidy3d/web/cli/develop/documentation.py:1]: Yeah exactly it was getting too large so I broke it up. I really need to move the CLI out of there but needs to be tested very thoroughly as to not affect the web api.

File: tidy3d/components/types_extra.py
Comment (daquinteroflex) [reply tidy3d/components/types_extra.py:1]: So I think if we add it to the top level module, then it's no longer on-the-fly importing, since these types are only used when a function imports them - that's why I did not add it in. 

The idea was that we have sections of code that are not part of the core but can be called within the core functions - that's how I we got the import speedups. What do you think?

File: tidy3d/components/geometry/polyslab.py
Comment (daquinteroflex) [reply tidy3d/components/geometry/polyslab.py:556]: This is actually really interesting - because I've been thinking about how best to do this in tidy3d 3.0 too, and it doesn't involve on-the-fly imports not necessarily like this format style.

There has been some debate around this here https://stackoverflow.com/questions/128478/should-import-statements-always-be-at-the-top-of-a-module and I think in our case it's fine. Re -importing the module within the function is still fine to solve the namespace situation and probably cleaner than the alternative variable as the memory of the imported library is managed by importlib rather than us

File: tidy3d/components/geometry/base.py
Comment (daquinteroflex) [reply tidy3d/components/geometry/base.py:1070]: Yeah it's a bit more confusing I agree.

So the idea was to consolidate the way we do lazy imports into a single methodology with the goal of optimizing that single function for timing performance.

The objective of this implementation was to let importlib sort out the namespace for us since that way we don't have to track library state variables so much. However I could, for example, refactor this to load the library using the `importlib.import_module` method which returns the library and this would remove the need for the `try-except` statements. The idea was to implement the same logic without having to manage the library state through ancilliary variables. However, can improve this implementation if we want to make it cleaner - but it again comes down to how do we do this best when having importlib manage the right library state. 

Let me see if there's other ways to do this (and also relevant for arch 3.0)

File: .github/workflows/test_develop_cli.yaml
Comment (tylerflex) [reply .github/workflows/test_develop_cli.yaml:10]: seems like 80% of the time we're developing on a pre/* branch, so probably we should be running on those.

What if we tested after a new commit is added to the pre/* branch (so AFTER a PR is merged?) or is this looking for trouble?

I think on release it might be too late.. we'd just have to patch a new release right after if it breaks.

File: docs/development/documentation.rst
Comment (tylerflex) [reply docs/development/documentation.rst:49]: I see, thanks!

File: tidy3d/components/geometry/base.py
Comment (tylerflex) [reply tidy3d/components/geometry/base.py:1070]: the way I read it is the difference between 1 and 2 below.

```py
def foo(x):
    have_package_a = check_module_import("A")
    if have_package_a and isinstance(x, TypeA):
        import A
        A.do_something(x)
    have_package_b = check_module_import("B")
    if have_package_b and isinstance(x, TypeB):
        import B
        B.do_something(x)
```

vs.

```py
def foo(x):
    if isinstance(x, TypeA):
        try:
            import A
            A.do_something(x)
        except ImportError:
            ...
            
    if isinstance(x, TypeB):
        try:
            import B
            B.do_something(x)
        except ImportError:
            ...
```
I find 1 to be easier to reason about and avoids try/except. I think the functionality / logic is the same, just that the try except is hidden in the "check_module_import()" in 1? 

> So the idea was to consolidate the way we do lazy imports into a single methodology with the goal of optimizing that single function for timing performance.

could you explain what this single methodology would look like in the context of `foo` above? I'm having trouble seeing it.

File: tidy3d/components/geometry/polyslab.py
Comment (tylerflex) [reply tidy3d/components/geometry/polyslab.py:556]: if the import has already been imported in another method, the 2nd import only takes minimal amount of time right? (unless we manually reload it with one of import lib's functions?)

I feel like fundamentally the problem here is that we're trying to mix optional dependencies within the core components, when they should probably be plugins.. although that could hurt the user experience a bit.

File: tidy3d/components/geometry/base.py
Comment (daquinteroflex) [reply tidy3d/components/geometry/base.py:1070]: Yeah sounds good, it makes sense this way, had not thought about it in this implementation - and this will remove the `try-except` blocks so I'll restructure this.

File: tests/test_cli/full_test_develop.py
Comment (daquinteroflex) [reply tests/test_cli/full_test_develop.py:98]: Implemented now!

File: .github/workflows/test_develop_cli.yaml
Comment (daquinteroflex) [reply .github/workflows/test_develop_cli.yaml:10]: So the tricky thing is these develop tests will also test the documentation build which is several minutes long. If we don't mind that then I guess it'd be fine to run on pre/ anytime

File: tidy3d/components/geometry/polyslab.py
Comment (daquinteroflex) [reply tidy3d/components/geometry/polyslab.py:556]: That's the idea, it's not great but working within the current structure.

Yeah I agree, I've been trying to keep them separate in different files and only use methods from those files within any core methods if needed. But really our architecture should be modular like the plugins in this regard I think too.

File: tests/test_cli/full_test_develop.py
Comment (daquinteroflex) [reply tests/test_cli/full_test_develop.py:28]: So I was following the recommended Click testing methods which use this https://click.palletsprojects.com/en/8.1.x/testing/#file-system-isolation

We would have control of a tmp path if we wanted it looks like

File: tests/test_cli/full_test_develop.py
Comment (daquinteroflex) [reply tests/test_cli/full_test_develop.py:1]: I think I need to write a full pytest suite that tests everything. If I test the standard tests* I get a 91% coverage, bu if I test the coverage of this tests/test_cli/full_test_develop I get 34%.

I'll write a proper test suite in another PR I think but happy to do it in this if we want to fully verify.

File: tidy3d/components/geometry/base.py
Comment (daquinteroflex) [reply tidy3d/components/geometry/base.py:1070]: So I think maybe the try statements are essential because of the way the `TypeA` and `TypeB` is defined as `gdspy.Cell` or `gdstk.Cell` type which means that the import of the library has to be done beforehand.

```
gdstk_available = check_import("gdstk")
if isinstance(gds_cell, gdstk.Cell):
    import gdstk

    if isinstance(gds_cell, gdstk.Cell):
        gds_loader_fn = Geometry.load_gds_vertic
```

**Comment by daquinteroflex** [comment]
So I've replied to all comments and apart from the decisions we need to take that I reference in the comments, then this is ready for re review

File: tests/test_cli/full_test_develop.py
Comment (tylerflex) [reply tests/test_cli/full_test_develop.py:1]: is this 91% coverage including the develop commands? That's still pretty good.

I guess what I'm curious about is the coverage of the `web/cli/develop` directory.

We can do this in another PR though.

File: tidy3d/components/geometry/base.py
Comment (tylerflex) [reply tidy3d/components/geometry/base.py:1070]: oh right, ok feel free to revert.

**Comment by tylerflex** [comment]
Looks good to me. Maybe we can have @lucas-flexcompute take a quick look through, especially on the optional import stuff?

**Comment by daquinteroflex** [comment]
So I think Lucas already reviewed the base 2.5 PR in which this was built https://github.com/flexcompute/tidy3d/pull/1376 and I implemented the changes both there which propagated here. I'm going to rebase to fix the conflict warning but apart from that maybe fine?

File: tidy3d/components/geometry/base.py
Comment (lucas-flexcompute) [reply tidy3d/components/geometry/base.py:1070]: If you import and test the type only after checking for availability, you can still skip the try/except, can't you?
```python
gdstk_available = check_import("gdstk")
if gdstk_available:
    import gdstk

    if isinstance(gds_cell, gdstk.Cell):
        gds_loader_fn = ...
```

Comment (daquinteroflex) [reply tidy3d/components/geometry/base.py:1070]: Sure sounds good. It may even be closer to the original logic just that on a local functional level. 

I guess my mind was a bit mentally blocked on this. Will restructure now to remove the try-excepts this way and it's cleaner to read. 

```
gdstk_available = check_import("gdstk")
if gdstk_available:
    import gdstk

    if isinstance(gds_cell, gdstk.Cell):
        gds_loader_fn = ...
        
gdspy_available = check_import("gdspy")
if gdspy_available:
    import gdspy

    if isinstance(gds_cell, gdspy.Cell):
        gds_loader_fn = ...

if not gdstk_available and gdspy_available:
    ...
```

File: tidy3d/components/geometry/polyslab.py
Comment (lucas-flexcompute) [reply tidy3d/components/geometry/polyslab.py:556]: One possible solution to avoid calling `import` within the function is to get the imported module from `sys.modules` (it might be more efficient as well, since we know the module is already imported):
```python
In [1]: import importlib

In [2]: import sys

In [3]: def check_import(s):
   ...:     try:
   ...:         importlib.import_module(s)
   ...:         return True
   ...:     except ImportError:
   ...:         return False
   ...:

In [4]: def f():
   ...:     if not check_import("numpy"):
   ...:         raise ImportError("numpy not available")
   ...:     # avoid import here
   ...:     numpy = sys.modules["numpy"]
   ...:     return numpy.arange(3)
   ...:

In [5]: f()
Out[5]: array([0, 1, 2])
```

Comment (daquinteroflex) [reply tidy3d/components/geometry/polyslab.py:556]: I was interested about this, turns out this is exactly what happens under the hood when we do `import_module`:

https://github.com/python/cpython/blob/c63c6142f9146e1e977f4c824c56e8979e6aca87/Lib/importlib/_bootstrap.py#L1354

**Comment by daquinteroflex** [comment]
Implemented Lucas's suggestions, basically rolled back the logic just that made it functionally local so should be about the same. Let me know if anything else.

**Review by tylerflex — APPROVED** [review]
looks good to me! can you just add some module level docstrings to the `develop/` modules?

**Comment by daquinteroflex** [comment]
Done, I'll rebase into pre/2.6 when the tests pass if that sounds good

**Pull request merged**
